### PR TITLE
Task #124742 feat : New feature - Auto save form

### PR DESCRIPTION
**Pull request by vaibhavsTekdi**
- Auto Save feature works on ‘onblur’ input trigger for every field.
- It saves the form in draft mode & it will work for darft record only. 
- As auto save saves the form in draft mode, we have removed "Allowed As Draft" param from admin panel & save draft button wont be anymore on site panel.
- If any user tries to navigate away from the form with unsaved fields then the alert should be shown. This is irrespective if the autosave is active or not. 

**Note :** After final submit you can edit the form. But autosave functionality will be disabled on edit form.

Regarding the UX concern while editing the form, we are alerting the user about unsaved changes before closing the page or refreshing the page.

### CODE REVIEW

File: administrator/assets/js/ays-beforeunload-shim.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+$(function() {
```
Comment (ankush-maherwal) [inline administrator/assets/js/ays-beforeunload-shim.js]: JS should be in UCM

File: administrator/assets/js/jquery.are-you-sure.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
+(function($) {
```
Comment (ankush-maherwal) [inline administrator/assets/js/jquery.are-you-sure.js]: JS should be in UCM

File: site/helpers/tjfields.php [hunk 3]
```diff
@@ -298,16 +298,13 @@ public function saveFieldsValue($data)
 			$conditions = array(
 				$db->quoteName('field_id') . ' = ' . $unsubmittedField,
 				$db->quoteName('content_id') . ' = ' . $data['content_id'],
-				$db->quoteName('client') . " = '" . $data['client'] . "'"
+				$db->quoteName('client') . " = " . $db->quote($data['client']) . ""
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: $db->quoteName('client') . " = " . $db->quote($data['client']) . ""
=>
$db->quoteName('client') . " = " . $db->quote($data['client'])

File: site/helpers/tjfields.php [hunk 4]
```diff
@@ -711,7 +708,7 @@ public function deleteFieldValueEntry($fieldValueEntryId)
 
 			// Delete all custom keys for user 1001.
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: extra comment

File: site/helpers/tjfields.php [hunk 5]
```diff
@@ -842,9 +841,10 @@ public function getCategorys($client)
 	{
 		$db    = JFactory::getDbo();
 		$query = $db->getQuery(true);
-		$query->select('* FROM #__categories');
-		$query->where('extension="' . $client . '"');
-		$query->where('published=1');
+		$query->select('*');
+		$query->from($db->quoteName('#__categories'));
```
Comment (ankush-maherwal) [inline site/helpers/tjfields.php]: can we use categories model to get list of categories?

Comment (vaibhavsTekdi) [reply site/helpers/tjfields.php]: Have added the tracker list : http://tracker.tekdi.net/issues/126669

**Pull request merged**